ConfigurationSystem/Agent/VOMS2CSAgent.py

```diff
@@ -75,37 +83,143 @@ def execute( self ):
+
+      if self.voChanged or self.detailedReport:
         mailMsg = ""
         if self.__adminMsgs[ 'Errors' ]:
           mailMsg += "\nErrors list:\n  %s" % "\n  ".join( self.__adminMsgs[ 'Errors' ] )
         if self.__adminMsgs[ 'Info' ]:
           mailMsg += "\nRun result:\n  %s" % "\n  ".join( self.__adminMsgs[ 'Info' ] )
+        if self.detailedReport:
+          result = self.getVOUserReport( vo )
+          if result['OK']:
+            mailMsg += '\n\n'
+            mailMsg += result['Value']
+          else:
+            mailMsg += 'Failed to produce a detailed user report'
+            mailMsg += result['Message']
         NotificationClient().sendMail( self.am_getOption( 'MailTo', voAdminMail ),
                                        "VOMS2CSAgent run log", mailMsg,
-                                       self.am_getOption( 'mailFrom', "DIRAC system" ) )
-
-    if self.csapi.csModified:
-      # We have accumulated all the changes, commit them now
-      self.log.info( "There are changes to the CS ready to be committed" )
-      if self.dryRun:
-        self.log.info( "Dry Run: CS won't be updated" )
-        self.csapi.showDiff()
-      else:
-        result = self.csapi.commitChanges()
-        if not result[ 'OK' ]:
-          self.log.error( "Could not commit configuration changes", result[ 'Message' ] )
-          return result
-        self.log.notice( "Configuration committed" )
-    else:
-      self.log.info( "No changes to the CS recorded at this cycle" )
+                                       self.am_getOption( 'MailFrom', self.am_getOption( 'mailFrom', "DIRAC system" ) ) )
 
     return S_OK()
 
+  def getVOUserData( self, vo, refreshFlag = False ):
+    """ Get a report for users of a given VO
+
+    :param str vo: VO name
+    :return: S_OK/S_ERROR, Value = user description dictionary
+    """
+    if refreshFlag:
+      gConfig.forceRefresh()
+
+    # Get DIRAC users
+    diracUsers = getUsersInVO( vo )
+    if not diracUsers:
+      return S_ERROR( "No VO users found for %s" % vo )
+
+    if refreshFlag:
+      result = self.csapi.downloadCSData()
+      if not result['OK']:
+        return result
+    result = self.csapi.describeUsers( diracUsers )
+    if not result['OK']:
+      self.log.error( 'Could not retrieve CS User description' )
+    return result
+
+  def getVOUserReport( self, vo ):
+    """
+
+    :param str vo: VO name
+    :return: report string
+    """
+
+    result = self.getVOUserData( vo, refreshFlag = True )
+    if not result['OK']:
+      return result
+
+    userDict = result['Value']
+
+    # Get DIRAC group vs VOMS Role Mappings
+    result = getVOMSRoleGroupMapping( vo )
+    if not result['OK']:
+      return result
+
+    diracVOMSMapping = result['Value']['DIRACVOMS']
+
+    records = []
+    groupDict = {}
+    multiDNUsers = {}
+    suspendedUsers = []
+    for user in userDict:
+      for group in userDict[user]['Groups']:
+        groupDict.setdefault( group, 0 )
```

Comment:
One could also use `from collections import defaultdict` for groupDict
https://docs.python.org/2/library/collections.html#collections.defaultdict